scrapers/scraper.py

```diff
@@ -0,0 +1,267 @@
+                    print(email)
+
+        shutil.rmtree("./apertium-{}/".format(language), onerror=del_rw)
+
+        authors = lines.split("\n")
+        authors.pop() #last line is always empty
+        contributors = []
+        authorCount = Counter(authors)
+        for contributor, count in authorCount.items():
+            contributors.append({"user": contributor, "value": count})
+
+        wikiData = requests.get(wikiURL).text
+        rows = etree.fromstring(wikiData, parser=etree.XMLParser(encoding="utf-8"))[1][0][0][0][3][3][8]
+        for row in rows[2:]:
+            if row[0][0][0].text[len("apertium-"):] == language:
+                state = row[6].text.strip()
+                break
+            state = "unknown"
+
+        data.append({
+            "lang": language,
+            "state": state,
+            "stems": stems,
+            "location": "apertium-{} ({})".format(language, location),
+            "contributors": contributors
+        })
+
+    return data
+
+
+def pairData(packages):
+    """Returns the locations and stems of all specified pairs"""
+    data = []
+    for package in packages:
+        if not re.match(r"apertium-\w+-\w+", package["name"]):
+            continue
+
+        pairName = package["name"].replace("apertium-", "")
+        pairSet = set(pairName.split("-"))
+        if not pairSet <= set(languages):
+            continue
+        for topic in package["topics"]:
+            if topic[len("apertium-"):] in pairLocations:
+                location = topic[len("apertium-"):]
+                break
+        stems = json.loads(requests.get("https://apertium.projectjj.com/stats-service/apertium-{}/bidix".format(pairName)).text)["stats"][0]["value"]
+        data.append({"langs": list(pairSet), "location": location, "stems": stems})
+
+    return data
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Scrape data for visualizer")
+    parser.add_argument("-s", "--shallow", help="Faster mode, doesn't dig through histories", action="store_true")
+    parser.add_argument("-u", "--updatemailmap", help="outputs users that aren't on .mailmap", action="store_true")
+    args = parser.parse_args()
+
+    logging.disable(logging.CRITICAL)
+
+    allPackages = json.loads(requests.get("https://apertium.projectjj.com/stats-service/packages").text)["packages"]
+    pairsFile = open("../json/pairData.json", "w+", encoding="utf-8")
+    json.dump(pairData(allPackages), pairsFile, ensure_ascii=False)
+    langsFile = open("../json/transducers.json", "w+", encoding="utf-8")
+    json.dump(monoData(allPackages, args.updatemailmap), langsFile, ensure_ascii=False)
+    for lang in languages:
+        if not args.shallow:
+            langHistory = []
+            langHistory.append(monoHistory(lang))
+            langHistory.extend(pairHistory(lang, allPackages))
+            outputFile = open("../json/{}.json".format(lang), "w+", encoding="utf-8")
```

Comment:
Do we really want to append to the files here and above?

Reply:
Why not?

Reply:
Because JSON appended to JSON isn't valid JSON?

Reply:
Ah, but this isn't appending. It's overwriting each file.

Reply:
Gotcha. I mixed up my file modes.